core/trino-main/src/main/java/io/trino/operator/output/RowPositionsAppender.java

```diff
@@ -0,0 +1,233 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.operator.output;
+
+import io.trino.spi.block.Block;
+import io.trino.spi.block.RowBlock;
+import io.trino.spi.block.RunLengthEncodedBlock;
+import io.trino.spi.type.RowType;
+import it.unimi.dsi.fastutil.ints.IntArrayList;
+import org.openjdk.jol.info.ClassLayout;
+
+import java.util.Arrays;
+import java.util.List;
+
+import static io.airlift.slice.SizeOf.sizeOf;
+import static io.trino.operator.output.PositionsAppenderUtil.calculateBlockResetSize;
+import static io.trino.operator.output.PositionsAppenderUtil.calculateNewArraySize;
+import static io.trino.spi.block.RowBlock.createRowBlockInternal;
+import static java.util.Objects.requireNonNull;
+
+public class RowPositionsAppender
+        implements PositionsAppender
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(RowPositionsAppender.class).instanceSize();
+    private final PositionsAppender[] fieldAppenders;
+    private int initialEntryCount;
+    private boolean initialized;
+
+    private int positionCount;
+    private boolean hasNullRow;
+    private boolean hasNonNullRow;
+    private boolean[] rowIsNull = new boolean[0];
+    private int[] fieldBlockOffsets = new int[1];
+    private long retainedSizeInBytes;
+    private long sizeInBytes;
+
+    public static RowPositionsAppender createRowAppender(
+            PositionsAppenderFactory positionsAppenderFactory,
+            RowType type,
+            int expectedPositions,
+            long maxPageSizeInBytes)
+    {
+        PositionsAppender[] fields = new PositionsAppender[type.getFields().size()];
+        for (int i = 0; i < fields.length; i++) {
+            fields[i] = positionsAppenderFactory.create(type.getFields().get(i).getType(), expectedPositions, maxPageSizeInBytes);
+        }
+        return new RowPositionsAppender(fields, expectedPositions);
+    }
+
+    private RowPositionsAppender(PositionsAppender[] fieldAppenders, int expectedPositions)
+    {
+        this.fieldAppenders = requireNonNull(fieldAppenders, "fields is null");
+        this.initialEntryCount = expectedPositions;
+        updateRetainedSize();
+    }
+
+    @Override
+    public void append(IntArrayList positions, Block block)
+    {
+        if (positions.isEmpty()) {
+            return;
+        }
+        ensureCapacity(positions.size());
+        RowBlock sourceRowBlock = (RowBlock) block;
+        IntArrayList notNullPositions;
+        if (sourceRowBlock.mayHaveNull()) {
+            notNullPositions = processNullablePositions(positions, sourceRowBlock);
+            hasNullRow |= notNullPositions.size() < positions.size();
+            hasNonNullRow |= notNullPositions.size() > 0;
+        }
+        else {
+            // the source Block does not have nulls
+            notNullPositions = new IntArrayList(positions.size());
```

Comment:
For consistency with `processNullablePositions`, we should start with raw array and use `IntArrayList.wrap` here too and maybe consider adding `processNonNullablePositions` method